pontoon/machinery/utils.py

```diff
@@ -96,37 +101,122 @@ def get_google_automl_translation(text, locale, format="text"):
         }
 
-    model_id = locale.google_automl_model
-
     # Google AutoML Translation requires location "us-central1"
     location = "us-central1"
 
     parent = f"projects/{project_id}/locations/{location}"
+    model_id = locale.google_automl_model
     model_path = f"{parent}/models/{model_id}"
 
-    response = client.translate_text(
-        request={
-            "contents": [text],
-            "target_language_code": locale.google_translate_code,
-            "model": model_path,
-            "source_language_code": "en",
-            "parent": parent,
-            "mime_type": "text/html" if format == "html" else "text/plain",
-        }
-    )
+    request_params = {
+        "contents": [text],
+        "target_language_code": locale.google_translate_code,
+        "model": model_path,
+        "source_language_code": "en",
+        "parent": parent,
+        "mime_type": "text/html" if format == "html" else "text/plain",
+    }
+
+    use_placeables_glossary(text, client, project_id, location, request_params)
+
+    # Get translations
+    response = client.translate_text(request=request_params)
 
-    if len(response.translations) == 0:
+    translations = response.translations
+    if response.glossary_translations:
+        translations = response.glossary_translations
+
+    if len(translations) == 0:
         return {
             "status": False,
             "message": "No translations found.",
         }
     else:
         return {
             "status": True,
-            "translation": response.translations[0].translated_text,
+            "translation": translations[0].translated_text,
         }
 
 
+def use_placeables_glossary(text, client, project_id, location, request_params):
+    placeables = get_placeables(text)
+
+    if not placeables:
+        return
+
+    glossary_id = "placeables"
+    glossary_path = client.glossary_path(project_id, location, glossary_id)
+    url = f"https://translation.googleapis.com/v3/{glossary_path}/glossaryEntries"
+
+    auth_req = google.auth.transport.requests.Request()
+    credentials = service_account.Credentials.from_service_account_file(
+        os.environ.get("GOOGLE_APPLICATION_CREDENTIALS"),
+        scopes=["https://www.googleapis.com/auth/cloud-platform"],
+    )
+    credentials.refresh(auth_req)
+    headers = {
+        "Authorization": f"Bearer {credentials.token}",
+        "Content-Type": "application/json; charset=utf-8",
+    }
+
+    # Retrieve already stored glossary terms
+    existing_terms = get_existing_terms(url, headers)
+
+    # Store any new terms to the glossary
+    if existing_terms:
+        new_terms = {term for term in placeables if term not in existing_terms}
+
+        if new_terms:
+            store_new_terms(url, headers, new_terms)
+
+    glossary_config = translate.TranslateTextGlossaryConfig(glossary=glossary_path)
+    request_params["glossary_config"] = glossary_config
+
+
+def get_existing_terms(url, headers):
```

Comment:
This seems like a thing that really ought to be cached. If that's too much to add to this PR, then please add a separate issue for it and a code comment here pointing at it.

Probably a new database table that we trust as the local source of truth that's updated during `store_new_terms()` and sync'd daily or weekly from upstream to account for any drift.